File: Venkatesh/Binary-Search-Tree/BST.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Linked_List:
    def __init__(self):
        self.head = None

    def insertion(self, data):
        if self.head is None:
            self.head = Node(data)
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = Node(data)

def convert_to_bst(linked_list):
    if linked_list.head is None:
        return []

    temp = linked_list.head
    arr = []
    while temp:
        arr.append(temp.data)
        temp = temp.next

    def build_bst(arr):
        if not arr:
            return None

        mid = len(arr) // 2
        root = Node(arr[mid])
        root.left = build_bst(arr[:mid])
        root.right = build_bst(arr[mid+1:])
        return root

    bst = build_bst(arr)

    def preorder(root):
        if root is None:
            return []

        result = []
        result.append(root.data)
        result.extend(preorder(root.left))
        result.extend(preorder(root.right))
        return result

    return preorder(bst)

linked_list = Linked_List()
linked_list.insertion(1)
linked_list.insertion(2)
linked_list.insertion(3)
linked_list.insertion(4)
linked_list.insertion(5)
linked_list.insertion(6)
linked_list.insertion(7)

print(convert_to_bst(linked_list))

File: Venkatesh/Binary-Search-Tree/BST.java
public class BST {

    Node head;


    class Node {
        int data;
        Node next;

        Node (int data) {
            this.data = data;
            this.next = null;
        }
    }

    static class TreeRoot {
        TreeRoot left;
        int data;
        TreeRoot right;
    
        TreeRoot (int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public void insertion(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node last = head;
            while (last.next != null) {
                last = last.next;
            }
            last.next = newNode;
        }
    }

    public static int[] convertToBST(Node head) {
        if (head == null) {
            return new int[0];
        }
           
        int count = 0;
        Node temp = head;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
           
        int[] preorder = new int[count];
           
        int[] arr = new int[count];
        temp = head;
        for (int i = 0; i < count; i++) {
            arr[i] = temp.data;
            temp = temp.next;
        }
           
        TreeRoot root = convertToBST(arr, 0, count - 1);
        
        preorderTraversal(root, preorder, 0);
        
        return preorder;
    }

    private static TreeRoot convertToBST(int[] arr, int start, int end) {
        if (start > end) {
            return null;
        }
        
        int mid = (start + end) / 2;
        
        TreeRoot root = new TreeRoot(arr[mid]);
        
        root.left = convertToBST(arr, start, mid - 1);
        root.right = convertToBST(arr, mid + 1, end);
        
        return root;
    }

    private static int preorderTraversal(TreeRoot TreeRoot, int[] preorder, int index) {
        if (TreeRoot == null) {
            return index;
        }
        preorder[index] = TreeRoot.data;
        index++;
        
        index = preorderTraversal(TreeRoot.left, preorder, index);
        index = preorderTraversal(TreeRoot.right, preorder, index);
        
        return index;
    }
    
    public static void main(String[] args) {
        BST list = new BST();
        list.insertion(1);
        list.insertion(2);
        list.insertion(3);
        list.insertion(4);
        list.insertion(5);
        list.insertion(6);
        list.insertion(7);

        int[] preorder = convertToBST(list.head);

        for (int i = 0; i < preorder.length; i++) {
            System.out.print(preorder[i] + " ");
        }
    }
}